File: Calculadora/EvalVisitor.java
import java.util.HashMap;
import java.util.Map;

public class EvalVisitor extends LabeledExprBaseVisitor<Integer> {
    Map<String, Integer> memory = new HashMap<>();

    @Override
    public Integer visitAssign(LabeledExprParser.AssignContext ctx) {
        String id = ctx.ID().getText();
        int value = visit(ctx.expr());
        memory.put(id, value);
        return value;
    }

    @Override
    public Integer visitPrintExpr(LabeledExprParser.PrintExprContext ctx) {
        Integer value = visit(ctx.expr());
        System.out.println(value);
        return 0;
    }

    @Override
    public Integer visitInt(LabeledExprParser.IntContext ctx) {
        return Integer.valueOf(ctx.INT().getText());
    }

    @Override
    public Integer visitId(LabeledExprParser.IdContext ctx) {
        String id = ctx.ID().getText();
        return memory.getOrDefault(id, 0);
    }

    @Override
    public Integer visitParens(LabeledExprParser.ParensContext ctx) {
        return visit(ctx.expr());
    }

    @Override
    public Integer visitMulDiv(LabeledExprParser.MulDivContext ctx) {
        int left = visit(ctx.expr(0));
        int right = visit(ctx.expr(1));

        return switch (ctx.op.getType()) {
            case LabeledExprParser.MUL -> left * right;
            case LabeledExprParser.DIV -> left / right;
            case LabeledExprParser.MOD -> left % right;
            case LabeledExprParser.POW -> (int) Math.pow(left, right);
            default -> 0;
        };
    }

    @Override
    public Integer visitAddSub(LabeledExprParser.AddSubContext ctx) {
        int left = visit(ctx.expr(0));
        int right = visit(ctx.expr(1));

        return switch (ctx.op.getType()) {
            case LabeledExprParser.ADD -> left + right;
            case LabeledExprParser.SUB -> left - right;
            default -> 0;
        };
    }

    @Override
    public Integer visitFuncCall(LabeledExprParser.FuncCallContext ctx) {
        String funcName = ctx.func.getText().toLowerCase();
        int value = visit(ctx.expr());

        return switch (funcName) {
            case "sin" -> (int) Math.round(Math.sin(Math.toRadians(value)));
            case "cos" -> (int) Math.round(Math.cos(Math.toRadians(value)));
            case "tan" -> (int) Math.round(Math.tan(Math.toRadians(value)));
            case "asin" -> (int) Math.round(Math.toDegrees(Math.asin(value)));
            case "acos" -> (int) Math.round(Math.toDegrees(Math.acos(value)));
            case "atan" -> (int) Math.round(Math.toDegrees(Math.atan(value)));
            default -> {
                System.err.println("Función desconocida: " + funcName);
                yield 0;
            }
        };
    }
}

File: Calculadora_en_python/EvalVisitor.py
from LabeledExprVisitor import LabeledExprVisitor
from LabeledExprParser import LabeledExprParser
import math

class EvalVisitor(LabeledExprVisitor):
    def __init__(self):
        self.memory = {}

    # --------- STATEMENTS ----------
    def visitAssign(self, ctx:LabeledExprParser.AssignContext):
        id = ctx.ID().getText()
        value = self.visit(ctx.expr())
        self.memory[id] = value
        return value

    def visitPrintExpr(self, ctx:LabeledExprParser.PrintExprContext):
        value = self.visit(ctx.expr())
        print(value)
        return value

    def visitBlank(self, ctx:LabeledExprParser.BlankContext):
        return None

    # --------- ATOMS ----------
    def visitInt(self, ctx:LabeledExprParser.IntContext):
        return int(ctx.INT().getText())

    def visitDouble(self, ctx:LabeledExprParser.DoubleContext):
        return float(ctx.DOUBLE().getText())

    def visitId(self, ctx:LabeledExprParser.IdContext):
        id = ctx.ID().getText()
        return self.memory.get(id, 0)

    def visitParens(self, ctx:LabeledExprParser.ParensContext):
        return self.visit(ctx.expr())

    # --------- OPERACIONES ----------
    def visitUnaryMinus(self, ctx:LabeledExprParser.UnaryMinusContext):
        return -self.visit(ctx.expr())

    def visitUnaryPlus(self, ctx:LabeledExprParser.UnaryPlusContext):
        return +self.visit(ctx.expr())

    def visitAddSub(self, ctx:LabeledExprParser.AddSubContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == LabeledExprParser.ADD:
            return left + right
        return left - right

    def visitMulDiv(self, ctx:LabeledExprParser.MulDivContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == LabeledExprParser.MUL:
            return left * right
        elif ctx.op.type == LabeledExprParser.DIV:
            return left / right
        elif ctx.op.type == LabeledExprParser.MOD:
            return left % right
        elif ctx.op.type == LabeledExprParser.POW:
            return left ** right
        else:
            return None

    def visitFactf(self, ctx:LabeledExprParser.FactfContext):
        value = self.visit(ctx.expr())
        return math.factorial(int(value))

    # --------- FUNCIONES ----------
    def visitFuncCall(self, ctx:LabeledExprParser.FuncCallContext):
        func = ctx.func.text.lower()
        value = self.visit(ctx.expr())
        if func == "sin":
            return math.sin(math.radians(value))
        elif func == "cos":
            return math.cos(math.radians(value))
        elif func == "tan":
            return math.tan(math.radians(value))
        else:
            print("Función desconocida:", func)
            return 0

    def visitFuncRad(self, ctx:LabeledExprParser.FuncRadContext):
        func = ctx.func.text.lower()
        value = self.visit(ctx.expr())
        if func == "sin":
            return math.sin(value)   # aquí ya se interpreta en radianes
        elif func == "cos":
            return math.cos(value)
        elif func == "tan":
            return math.tan(value)
        else:
            print("Función desconocida con rad:", func)
            return 0

    def visitSqrtf(self, ctx:LabeledExprParser.SqrtfContext):
        if self.visit(ctx.expr()) < 0:
            print("Error: Raíz cuadrada de un número negativo")
            return 0
        return math.sqrt(self.visit(ctx.expr()))

    def visitLnf(self, ctx:LabeledExprParser.LnfContext):
        return math.log(self.visit(ctx.expr()))

    def visitLogf(self, ctx:LabeledExprParser.LogfContext):
        return math.log10(self.visit(ctx.expr()))

    # --------- PROGRAMA ----------
    def visitProg(self, ctx:LabeledExprParser.ProgContext):
        result = None
        for child in ctx.getChildren():
            result = self.visit(child)
        return result
